server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -399,46 +398,88 @@ public void deleteSnapshot(SnapshotId snapshotId, long repositoryStateId, Action
-                    .map(info -> info.indices().stream().map(repositoryData::resolveIndexId).collect(Collectors.toList()))
+                    .map(info -> info.indices().stream().filter(survivingIndices::containsKey)
+                        .map(updatedRepositoryData::resolveIndexId).collect(Collectors.toList()))
                     .orElse(Collections.emptyList()),
                 snapshotId,
                 ActionListener.map(listener, v -> {
-                    try {
-                        blobStore().blobContainer(basePath().add("indices")).deleteBlobsIgnoringIfNotExists(
-                            unreferencedIndices.stream().map(IndexId::getId).collect(Collectors.toList()));
-                    } catch (IOException e) {
-                        logger.warn(() ->
-                            new ParameterizedMessage(
-                                "[{}] indices {} are no longer part of any snapshots in the repository, " +
-                                    "but failed to clean up their index folders.", metadata.name(), unreferencedIndices), e);
-                    }
+                    cleanupStaleIndices(indicesBlobContainer, foundIndices, survivingIndices);
                     return null;
                 })
             );
         }
     }
 
+    private void cleanupStaleIndices(BlobContainer indicesBlobContainer, Map<String, BlobContainer> foundIndices,
+                                     Map<String, IndexId> survivingIndices) {
+        try {
+            final Set<String> survivingIndexIds = survivingIndices.values().stream()
+                .map(IndexId::getId).collect(Collectors.toSet());
+            final List<String> toDelete = new ArrayList<>();
+            for (Map.Entry<String, BlobContainer> indexEntry : foundIndices.entrySet()) {
+                final String indexSnId = indexEntry.getKey();
+                try {
+                    if (survivingIndexIds.contains(indexSnId) == false) {
+                        logger.debug("[{}] Found stale index [{}]. Cleaning it up.", metadata.name(), indexSnId);
+                        deleteContents(indexEntry.getValue());
+                        toDelete.add(indexSnId);
+                        logger.debug("[{}] Cleaned up stale index [{}].", metadata.name(), indexSnId);
+                    }
+                } catch (IOException e) {
+                    logger.warn(() -> new ParameterizedMessage(
+                        "[{}] index {} is no longer part of any snapshots in the repository, " +
+                            "but failed to clean up their index folders.", metadata.name(), indexSnId), e);
+                }
+            }
+            try {
+                indicesBlobContainer.deleteBlobsIgnoringIfNotExists(toDelete);
+            } catch (IOException e) {
+                logger.warn(() ->
+                    new ParameterizedMessage(
+                        "[{}] failed to clean up unreferenced index folders.", metadata.name()), e);
```

Comment:
add folder names here as well?

Reply:
also, no need to add a full stop at the end of the log message (same for other messages you've added as well).